### DISCUSSION TITLE
Type-graphql cannot determine input type, even if I created a dto input file.

### DISCUSSION THREAD

**Discussion by SirMishaa** [MichalLytek/type-graphql#810] [General]
Hello, 
In my case, GraphQL input of class are useless because I'm creating **my own input** in my dto folder.

```ts
// -- dto/create-brand.input.ts
import { Field, InputType } from '@nestjs/graphql';
import { Product } from '@domain/product/model/product.entity';

@InputType()
export class CreateBrandInput {
  @Field(() => String)
  name!: string;

  @Field(() => [Product], { nullable: 'itemsAndList' })
  products?: Array<Product>;
}
```

But this does not seem to be enough, if the ``@InputType`` decorator is not also in the entity, **an error occurs**.

```ts
// -- brand.model.ts
@Entity()
@ObjectType()
@InputType('BrandInput')
export class Brand extends BaseEntity {  // .... Some fields }
```

The problem is that by putting it on, it creates an input with the **whole class** in the GraphQL schema (which is useless in my case). 

```
input CreateBrandInput {
  name: String!
  products: [ProductInput]
}

input BrandInput { 
  createdAt: String!
  id: String!
  name: String!
  products: [ProductInput]!
}
```
Here, ``BrandInput`` is useless, cause for creating a new Brand, I'm gonna use my own input, ``CreateBrandInput``. 

But If I'm removing the ``@InputType`` in the file ``brand.model.ts``, I've got an error (Error: Cannot determine a GraphQL input type for the "brand")

Am I doing something wrong? 
Thank you in advance.

**Comment by MichalLytek** [comment]
In GraphQL, input objects have a separate type in the system because object types can contain fields that express circular references or references to interfaces and unions, neither of which are appropriate for use as input arguments. 

That's why TypeGraphQL doesn't have a dedicated mechanism to reuse or convert object types to input types automatically.
You should keep them separated as they represent two different things in GraphQL. The `@InputType('BrandInput')` is just an unofficial workaround for the simplest cases and for developers who are aware of the limitation.

> Here, BrandInput is useless, cause for creating a new Brand, I'm gonna use my own input, CreateBrandInput.

You should put the `@InputType` decorator on the `Product` entity, not `Brand`, if you create your own input type.
And if you use the explicit `buildSchema -> resolvers` approach, all the not used types are omitted from the schema, so the `BrandInput` should not appear in the schema.

**Reply by SirMishaa** [reply]
But why does it need to convert object types to input types? ``Input Type`` is used only for mutation, no? 

Why can't I only have my own input type like this ?: 
```ts
// -- dto/create-brand.input.ts
import { Field, InputType } from '@nestjs/graphql';
import { Product } from '@domain/product/model/product.entity';

@InputType()
export class CreateBrandInput {
  @Field(() => String)
  name!: string;

  @Field(() => [Product], { nullable: 'itemsAndList' })
  products?: Array<Product>;
}
```

I'm using the code first approach, schema are automatically built using the GraphQLModule from `@nestjs/graphql` : 
```ts
GraphQLModule.forRoot({
      debug: true,
      playground: true,
      autoSchemaFile: path.join(process.cwd(), 'src/schema.gql'),
      sortSchema: true,
}),

export class AppModule {}
```

> The @InputType('BrandInput') is just an unofficial workaround for the simplest cases and for developers who are aware of the limitation.

In my case, I don't want it, is there any way to remove it using the code first approach? It look like I have to put ``@InputType`` at top of my entities since I have relation between them.

**Reply by MichalLytek** [reply]
> I'm using the code first approach, schema are automatically built using the GraphQLModule from @nestjs/graphql :

Please ask on a proper repo:
https://github.com/nestjs/graphql
NestJS has forked TypeGraphQL so I don't provide any support for NestJS GraphQL.
